reporter.go

```diff
@@ -39,3 +39,37 @@ type StatsReporter interface {
 	// Flush is expected to be called by a Scope when it completes a round or reporting
 	Flush()
 }
+
+// CachedStatsReporter is a backend for Scopes that pre allocates all
+// counter, gauges & timers. This is harder to implement but more performant
+type CachedStatsReporter interface {
+	// AllocateCounter pre allocates a counter data structure with name & tags.
+	AllocateCounter(name string, tags map[string]string) CachedCount
+
+	// AllocateGauge pre allocates a gauge data structure with name & tags.
+	AllocateGauge(name string, tags map[string]string) CachedGauge
+
+	// AllocateTimer pre allocates a timer data structure with name & tags.
+	AllocateTimer(name string, tags map[string]string) CachedTimer
+
+	// CachedCapabilities returns a description of metrics reporting capabilities
+	CachedCapabilities() Capabilities
+
+	// CachedFlush is expected to be called by a Scope when it completes a round or reporting
```

Comment:
Might be worth fixing the doc string here, the "or" is meant to be "of".  Worth fixing on line 39 too I should think.

I think it might be worth renaming `CachedCapabilities` and `CachedFlush` to just `Capabilities` and `Flush` and then creating a new internal only interface:

```go
type baseStatsReporter interface {
  Capabilities() Capabilities()
  Flush()
}
```

That way you can add a field to the scope:
```go
  reporter StatsReporter   // existing
  cachedReporter CachedStatsReporter // existing
  baseReporter baseStatsReporter // new
```

And when calling capabilities or flush:

```go
// capabilities
s.baseReporter.Capabilities()
// flush
s.baseReporter.Flush()
```

Just to avoid the if/elses there.